File: samples/language-impl-comparison/fsharp/json_extended.fs
/// JSON拡張版：コメント・トレーリングカンマ対応。
///
/// 標準JSONからの拡張点：
/// 1. コメント対応（`//` 行コメント、`/* */` ブロックコメント）
/// 2. トレーリングカンマ許可（配列・オブジェクトの最後の要素の後）
/// 3. より詳細なエラーメッセージ
///
/// 実用的な設定ファイル形式として：
/// - `package.json` 風の設定ファイル
/// - `.babelrc`, `.eslintrc` など開発ツールの設定
/// - VS Code の `settings.json`
module JsonExtended

open System
open System.Collections.Generic

// 型定義

type JsonValue =
  | JNull
  | JBool of bool
  | JNumber of float
  | JString of string
  | JArray of JsonValue list
  | JObject of Map<string, JsonValue>

type ParseError =
  | UnexpectedEOF
  | InvalidValue of string
  | UnclosedString
  | UnclosedBlockComment
  | ExpectedChar of char
  | InvalidNumber of string

type State =
  { Input: string
    Pos: int }

// パース

let parse (input: string) : Result<JsonValue, ParseError> =
  let rec skipWhitespaceAndComments (state: State) : Result<State, ParseError> =
    let skipWs (st: State) =
      let mutable pos = st.Pos
      while pos < st.Input.Length && (st.Input.[pos] = ' ' || st.Input.[pos] = '\n' || st.Input.[pos] = '\t' || st.Input.[pos] = '\r') do
        pos <- pos + 1
      { st with Pos = pos }

    let skipLineComment (st: State) =
      let newPos = st.Pos + 2
      let idx = st.Input.IndexOf('\n', newPos)
      let finalPos = if idx = -1 then st.Input.Length else idx + 1
      { st with Pos = finalPos }

    let skipBlockComment (st: State) : Result<State, ParseError> =
      let newPos = st.Pos + 2
      let idx = st.Input.IndexOf("*/", newPos)
      if idx = -1 then
        Error UnclosedBlockComment
      else
        Ok { st with Pos = idx + 2 }

    let stateAfterWs = skipWs state
    if stateAfterWs.Pos >= stateAfterWs.Input.Length then
      Ok stateAfterWs
    else if stateAfterWs.Pos + 1 < stateAfterWs.Input.Length && stateAfterWs.Input.Substring(stateAfterWs.Pos, 2) = "//" then
      skipWhitespaceAndComments (skipLineComment stateAfterWs)
    else if stateAfterWs.Pos + 1 < stateAfterWs.Input.Length && stateAfterWs.Input.Substring(stateAfterWs.Pos, 2) = "/*" then
      match skipBlockComment stateAfterWs with
      | Ok st -> skipWhitespaceAndComments st
      | Error e -> Error e
    else
      Ok stateAfterWs

  let rec parseValue (state: State) : Result<JsonValue * State, ParseError> =
    match skipWhitespaceAndComments state with
    | Error e -> Error e
    | Ok cleanState ->
      if cleanState.Pos >= cleanState.Input.Length then
        Error UnexpectedEOF
      else
        let remaining = cleanState.Input.Substring(cleanState.Pos)
        if remaining.StartsWith("null") then
          Ok (JNull, { cleanState with Pos = cleanState.Pos + 4 })
        else if remaining.StartsWith("true") then
          Ok (JBool true, { cleanState with Pos = cleanState.Pos + 4 })
        else if remaining.StartsWith("false") then
          Ok (JBool false, { cleanState with Pos = cleanState.Pos + 5 })
        else if remaining.StartsWith("\"") then
          parseString cleanState
        else if remaining.StartsWith("[") then
          parseArray cleanState
        else if remaining.StartsWith("{") then
          parseObject cleanState
        else
          parseNumber cleanState

  and parseString (state: State) : Result<JsonValue * State, ParseError> =
    let rec findEnd (pos: int) (acc: string) =
      if pos >= state.Input.Length then
        Error UnclosedString
      else
        match state.Input.[pos] with
        | '"' -> Ok (acc, pos + 1)
        | '\\' when pos + 1 < state.Input.Length ->
          let escaped =
            match state.Input.[pos + 1] with
            | 'n' -> "\n"
            | 't' -> "\t"
            | 'r' -> "\r"
            | '\\' -> "\\"
            | '"' -> "\""
            | ch -> string ch
          findEnd (pos + 2) (acc + escaped)
        | ch -> findEnd (pos + 1) (acc + string ch)

    match findEnd (state.Pos + 1) "" with
    | Ok (str, endPos) -> Ok (JString str, { state with Pos = endPos })
    | Error e -> Error e

  and parseNumber (state: State) : Result<JsonValue * State, ParseError> =
    let isNumChar ch =
      ch = '-' || ch = '+' || ch = '.' || ch = 'e' || ch = 'E' || (ch >= '0' && ch <= '9')

    let mutable endPos = state.Pos
    while endPos < state.Input.Length && isNumChar state.Input.[endPos] do
      endPos <- endPos + 1

    let numStr = state.Input.Substring(state.Pos, endPos - state.Pos)
    match Double.TryParse(numStr) with
    | (true, num) -> Ok (JNumber num, { state with Pos = endPos })
    | (false, _) -> Error (InvalidNumber numStr)

  and parseArray (state: State) : Result<JsonValue * State, ParseError> =
    let stateAfterBracket = { state with Pos = state.Pos + 1 }
    match skipWhitespaceAndComments stateAfterBracket with
    | Error e -> Error e
    | Ok cleanState ->
      if cleanState.Pos < cleanState.Input.Length && cleanState.Input.[cleanState.Pos] = ']' then
        Ok (JArray [], { cleanState with Pos = cleanState.Pos + 1 })
      else
        parseArrayElements cleanState []

  and parseArrayElements (state: State) (acc: JsonValue list) : Result<JsonValue * State, ParseError> =
    match parseValue state with
    | Error e -> Error e
    | Ok (value, stateAfterValue) ->
      match skipWhitespaceAndComments stateAfterValue with
      | Error e -> Error e
      | Ok cleanState ->
        let newAcc = acc @ [value]
        if cleanState.Pos >= cleanState.Input.Length then
          Error UnexpectedEOF
        else
          match cleanState.Input.[cleanState.Pos] with
          | ',' ->
            let stateAfterComma = { cleanState with Pos = cleanState.Pos + 1 }
            match skipWhitespaceAndComments stateAfterComma with
            | Error e -> Error e
            | Ok stateAfterWs ->
              if stateAfterWs.Pos < stateAfterWs.Input.Length && stateAfterWs.Input.[stateAfterWs.Pos] = ']' then
                // トレーリングカンマ
                Ok (JArray newAcc, { stateAfterWs with Pos = stateAfterWs.Pos + 1 })
              else
                parseArrayElements stateAfterWs newAcc
          | ']' ->
            Ok (JArray newAcc, { cleanState with Pos = cleanState.Pos + 1 })
          | _ ->
            Error (ExpectedChar ',')

  and parseObject (state: State) : Result<JsonValue * State, ParseError> =
    let stateAfterBrace = { state with Pos = state.Pos + 1 }
    match skipWhitespaceAndComments stateAfterBrace with
    | Error e -> Error e
    | Ok cleanState ->
      if cleanState.Pos < cleanState.Input.Length && cleanState.Input.[cleanState.Pos] = '}' then
        Ok (JObject Map.empty, { cleanState with Pos = cleanState.Pos + 1 })
      else
        parseObjectPairs cleanState Map.empty

  and parseObjectPairs (state: State) (acc: Map<string, JsonValue>) : Result<JsonValue * State, ParseError> =
    match parseString state with
    | Error e -> Error e
    | Ok (JString key, stateAfterKey) ->
      match skipWhitespaceAndComments stateAfterKey with
      | Error e -> Error e
      | Ok cleanState1 ->
        if cleanState1.Pos >= cleanState1.Input.Length || cleanState1.Input.[cleanState1.Pos] <> ':' then
          Error (ExpectedChar ':')
        else
          let stateAfterColon = { cleanState1 with Pos = cleanState1.Pos + 1 }
          match skipWhitespaceAndComments stateAfterColon with
          | Error e -> Error e
          | Ok cleanState2 ->
            match parseValue cleanState2 with
            | Error e -> Error e
            | Ok (value, stateAfterValue) ->
              match skipWhitespaceAndComments stateAfterValue with
              | Error e -> Error e
              | Ok cleanState3 ->
                let newAcc = Map.add key value acc
                if cleanState3.Pos >= cleanState3.Input.Length then
                  Error UnexpectedEOF
                else
                  match cleanState3.Input.[cleanState3.Pos] with
                  | ',' ->
                    let stateAfterComma = { cleanState3 with Pos = cleanState3.Pos + 1 }
                    match skipWhitespaceAndComments stateAfterComma with
                    | Error e -> Error e
                    | Ok stateAfterWs ->
                      if stateAfterWs.Pos < stateAfterWs.Input.Length && stateAfterWs.Input.[stateAfterWs.Pos] = '}' then
                        // トレーリングカンマ
                        Ok (JObject newAcc, { stateAfterWs with Pos = stateAfterWs.Pos + 1 })
                      else
                        parseObjectPairs stateAfterWs newAcc
                  | '}' ->
                    Ok (JObject newAcc, { cleanState3 with Pos = cleanState3.Pos + 1 })
                  | _ ->
                    Error (ExpectedChar ',')
    | Ok _ -> Error (InvalidValue "オブジェクトのキーは文字列である必要があります")
    | Error e -> Error e

  let initialState = { Input = input; Pos = 0 }
  match skipWhitespaceAndComments initialState with
  | Error e -> Error e
  | Ok st1 ->
    match parseValue st1 with
    | Error e -> Error e
    | Ok (value, st2) ->
      match skipWhitespaceAndComments st2 with
      | Error e -> Error e
      | Ok finalState ->
        if finalState.Pos >= finalState.Input.Length then
          Ok value
        else
          Error (InvalidValue "入力の終端に到達していません")

// レンダリング

let rec renderToString (value: JsonValue) (indentLevel: int) : string =
  let indent = String.replicate indentLevel "  "
  let nextIndent = String.replicate (indentLevel + 1) "  "

  match value with
  | JNull -> "null"
  | JBool true -> "true"
  | JBool false -> "false"
  | JNumber num -> num.ToString()
  | JString str -> "\"" + str + "\""
  | JArray items ->
    if List.isEmpty items then
      "[]"
    else
      let itemsStr =
        items
        |> List.map (fun item -> nextIndent + renderToString item (indentLevel + 1))
        |> String.concat ",\n"
      "[\n" + itemsStr + "\n" + indent + "]"
  | JObject pairs ->
    if Map.isEmpty pairs then
      "{}"
    else
      let pairsStr =
        pairs
        |> Map.toList
        |> List.map (fun (key, value) ->
          nextIndent + "\"" + key + "\": " + renderToString value (indentLevel + 1)
        )
        |> String.concat ",\n"
      "{\n" + pairsStr + "\n" + indent + "}"

// テスト

let testExtendedJson () =
  let testCases =
    [
      ("コメント対応", """
{
  // これは行コメント
  "name": "test",
  /* これは
     ブロックコメント */
  "version": "1.0"
}
""")
      ("トレーリングカンマ", """
{
  "items": [
    1,
    2,
    3,
  ],
  "config": {
    "debug": true,
    "port": 8080,
  }
}
""")
    ]

  testCases
  |> List.iter (fun (name, jsonStr) ->
    printfn "--- %s ---" name
    match parse jsonStr with
    | Ok value ->
      printfn "パース成功:"
      printfn "%s" (renderToString value 0)
    | Error err ->
      printfn "パースエラー: %A" err
    printfn ""
  )

File: samples/language-impl-comparison/ocaml/json_extended.ml
(** JSON拡張版：コメント・トレーリングカンマ対応。

    標準JSONからの拡張点：
    1. コメント対応（[//] 行コメント、[/* */] ブロックコメント）
    2. トレーリングカンマ許可（配列・オブジェクトの最後の要素の後）
    3. より詳細なエラーメッセージ

    実用的な設定ファイル形式として：
    - [package.json] 風の設定ファイル
    - [.babelrc], [.eslintrc] など開発ツールの設定
    - VS Code の [settings.json]
*)

(* 型定義 *)

type json_value =
  | JNull
  | JBool of bool
  | JNumber of float
  | JString of string
  | JArray of json_value list
  | JObject of (string * json_value) list

type parse_error =
  | UnexpectedEOF
  | InvalidValue of string
  | UnclosedString
  | UnclosedBlockComment
  | ExpectedChar of char
  | InvalidNumber of string

type state = {
  input : string;
  pos : int;
}

(* パース *)

let skip_ws state =
  let len = String.length state.input in
  let rec loop pos =
    if pos >= len then pos
    else
      match state.input.[pos] with
      | ' ' | '\n' | '\t' | '\r' -> loop (pos + 1)
      | _ -> pos
  in
  { state with pos = loop state.pos }

let skip_line_comment state =
  let len = String.length state.input in
  let new_pos = state.pos + 2 in
  let rec loop pos =
    if pos >= len then pos
    else if state.input.[pos] = '\n' then pos + 1
    else loop (pos + 1)
  in
  { state with pos = loop new_pos }

let skip_block_comment state =
  let len = String.length state.input in
  let new_pos = state.pos + 2 in
  let rec loop pos =
    if pos + 1 >= len then None
    else if state.input.[pos] = '*' && state.input.[pos + 1] = '/' then Some (pos + 2)
    else loop (pos + 1)
  in
  match loop new_pos with
  | None -> Error UnclosedBlockComment
  | Some end_pos -> Ok { state with pos = end_pos }

let rec skip_whitespace_and_comments state =
  let state_after_ws = skip_ws state in
  if state_after_ws.pos >= String.length state_after_ws.input then
    Ok state_after_ws
  else if state_after_ws.pos + 1 < String.length state_after_ws.input then
    let ch1 = state_after_ws.input.[state_after_ws.pos] in
    let ch2 = state_after_ws.input.[state_after_ws.pos + 1] in
    if ch1 = '/' && ch2 = '/' then
      skip_whitespace_and_comments (skip_line_comment state_after_ws)
    else if ch1 = '/' && ch2 = '*' then
      match skip_block_comment state_after_ws with
      | Error e -> Error e
      | Ok st -> skip_whitespace_and_comments st
    else
      Ok state_after_ws
  else
    Ok state_after_ws

let parse_string state =
  let len = String.length state.input in
  let rec loop pos acc =
    if pos >= len then Error UnclosedString
    else
      match state.input.[pos] with
      | '"' -> Ok (JString acc, { state with pos = pos + 1 })
      | '\\' when pos + 1 < len ->
          let escaped = match state.input.[pos + 1] with
            | 'n' -> "\n"
            | 't' -> "\t"
            | 'r' -> "\r"
            | '\\' -> "\\"
            | '"' -> "\""
            | ch -> String.make 1 ch
          in
          loop (pos + 2) (acc ^ escaped)
      | ch -> loop (pos + 1) (acc ^ String.make 1 ch)
  in
  loop (state.pos + 1) ""

let parse_number state =
  let len = String.length state.input in
  let is_num_char ch =
    ch = '-' || ch = '+' || ch = '.' || ch = 'e' || ch = 'E' ||
    (ch >= '0' && ch <= '9')
  in
  let rec loop pos =
    if pos >= len then pos
    else if is_num_char state.input.[pos] then loop (pos + 1)
    else pos
  in
  let end_pos = loop state.pos in
  let num_str = String.sub state.input state.pos (end_pos - state.pos) in
  try
    let num = float_of_string num_str in
    Ok (JNumber num, { state with pos = end_pos })
  with Failure _ -> Error (InvalidNumber num_str)

let rec parse_value state =
  match skip_whitespace_and_comments state with
  | Error e -> Error e
  | Ok clean_state ->
      if clean_state.pos >= String.length clean_state.input then
        Error UnexpectedEOF
      else
        let remaining = String.sub clean_state.input clean_state.pos
          (String.length clean_state.input - clean_state.pos) in
        if String.length remaining >= 4 && String.sub remaining 0 4 = "null" then
          Ok (JNull, { clean_state with pos = clean_state.pos + 4 })
        else if String.length remaining >= 4 && String.sub remaining 0 4 = "true" then
          Ok (JBool true, { clean_state with pos = clean_state.pos + 4 })
        else if String.length remaining >= 5 && String.sub remaining 0 5 = "false" then
          Ok (JBool false, { clean_state with pos = clean_state.pos + 5 })
        else if remaining.[0] = '"' then
          parse_string clean_state
        else if remaining.[0] = '[' then
          parse_array clean_state
        else if remaining.[0] = '{' then
          parse_object clean_state
        else
          parse_number clean_state

and parse_array state =
  let state_after_bracket = { state with pos = state.pos + 1 } in
  match skip_whitespace_and_comments state_after_bracket with
  | Error e -> Error e
  | Ok clean_state ->
      if clean_state.pos < String.length clean_state.input &&
         clean_state.input.[clean_state.pos] = ']' then
        Ok (JArray [], { clean_state with pos = clean_state.pos + 1 })
      else
        parse_array_elements clean_state []

and parse_array_elements state acc =
  match parse_value state with
  | Error e -> Error e
  | Ok (value, state_after_value) ->
      match skip_whitespace_and_comments state_after_value with
      | Error e -> Error e
      | Ok clean_state ->
          let new_acc = acc @ [value] in
          if clean_state.pos >= String.length clean_state.input then
            Error UnexpectedEOF
          else
            match clean_state.input.[clean_state.pos] with
            | ',' ->
                let state_after_comma = { clean_state with pos = clean_state.pos + 1 } in
                (match skip_whitespace_and_comments state_after_comma with
                | Error e -> Error e
                | Ok state_after_ws ->
                    if state_after_ws.pos < String.length state_after_ws.input &&
                       state_after_ws.input.[state_after_ws.pos] = ']' then
                      (* トレーリングカンマ *)
                      Ok (JArray new_acc, { state_after_ws with pos = state_after_ws.pos + 1 })
                    else
                      parse_array_elements state_after_ws new_acc)
            | ']' ->
                Ok (JArray new_acc, { clean_state with pos = clean_state.pos + 1 })
            | _ ->
                Error (ExpectedChar ',')

and parse_object state =
  let state_after_brace = { state with pos = state.pos + 1 } in
  match skip_whitespace_and_comments state_after_brace with
  | Error e -> Error e
  | Ok clean_state ->
      if clean_state.pos < String.length clean_state.input &&
         clean_state.input.[clean_state.pos] = '}' then
        Ok (JObject [], { clean_state with pos = clean_state.pos + 1 })
      else
        parse_object_pairs clean_state []

and parse_object_pairs state acc =
  match parse_string state with
  | Error e -> Error e
  | Ok (JString key, state_after_key) ->
      (match skip_whitespace_and_comments state_after_key with
      | Error e -> Error e
      | Ok clean_state1 ->
          if clean_state1.pos >= String.length clean_state1.input ||
             clean_state1.input.[clean_state1.pos] <> ':' then
            Error (ExpectedChar ':')
          else
            let state_after_colon = { clean_state1 with pos = clean_state1.pos + 1 } in
            (match skip_whitespace_and_comments state_after_colon with
            | Error e -> Error e
            | Ok clean_state2 ->
                (match parse_value clean_state2 with
                | Error e -> Error e
                | Ok (value, state_after_value) ->
                    (match skip_whitespace_and_comments state_after_value with
                    | Error e -> Error e
                    | Ok clean_state3 ->
                        let new_acc = acc @ [(key, value)] in
                        if clean_state3.pos >= String.length clean_state3.input then
                          Error UnexpectedEOF
                        else
                          match clean_state3.input.[clean_state3.pos] with
                          | ',' ->
                              let state_after_comma = { clean_state3 with pos = clean_state3.pos + 1 } in
                              (match skip_whitespace_and_comments state_after_comma with
                              | Error e -> Error e
                              | Ok state_after_ws ->
                                  if state_after_ws.pos < String.length state_after_ws.input &&
                                     state_after_ws.input.[state_after_ws.pos] = '}' then
                                    (* トレーリングカンマ *)
                                    Ok (JObject new_acc, { state_after_ws with pos = state_after_ws.pos + 1 })
                                  else
                                    parse_object_pairs state_after_ws new_acc)
                          | '}' ->
                              Ok (JObject new_acc, { clean_state3 with pos = clean_state3.pos + 1 })
                          | _ ->
                              Error (ExpectedChar ',')))))
  | Ok _ -> Error (InvalidValue "オブジェクトのキーは文字列である必要があります")
  | Error e -> Error e

let parse input =
  let initial_state = { input; pos = 0 } in
  match skip_whitespace_and_comments initial_state with
  | Error e -> Error e
  | Ok st1 ->
      match parse_value st1 with
      | Error e -> Error e
      | Ok (value, st2) ->
          match skip_whitespace_and_comments st2 with
          | Error e -> Error e
          | Ok final_state ->
              if final_state.pos >= String.length final_state.input then
                Ok value
              else
                Error (InvalidValue "入力の終端に到達していません")

(* レンダリング *)

let rec render_to_string value indent_level =
  let indent = String.make (indent_level * 2) ' ' in
  let next_indent = String.make ((indent_level + 1) * 2) ' ' in
  match value with
  | JNull -> "null"
  | JBool true -> "true"
  | JBool false -> "false"
  | JNumber num -> string_of_float num
  | JString str -> "\"" ^ str ^ "\""
  | JArray items ->
      if items = [] then "[]"
      else
        let items_str =
          items
          |> List.map (fun item -> next_indent ^ render_to_string item (indent_level + 1))
          |> String.concat ",\n"
        in
        "[\n" ^ items_str ^ "\n" ^ indent ^ "]"
  | JObject pairs ->
      if pairs = [] then "{}"
      else
        let pairs_str =
          pairs
          |> List.map (fun (key, value) ->
            next_indent ^ "\"" ^ key ^ "\": " ^ render_to_string value (indent_level + 1))
          |> String.concat ",\n"
        in
        "{\n" ^ pairs_str ^ "\n" ^ indent ^ "}"

(* テスト *)

let test_extended_json () =
  let test_cases = [
    ("コメント対応", {|
{
  // これは行コメント
  "name": "test",
  /* これは
     ブロックコメント */
  "version": "1.0"
}
|});
    ("トレーリングカンマ", {|
{
  "items": [
    1,
    2,
    3,
  ],
  "config": {
    "debug": true,
    "port": 8080,
  }
}
|})
  ] in
  List.iter (fun (name, json_str) ->
    Printf.printf "--- %s ---\n" name;
    match parse json_str with
    | Ok value ->
        Printf.printf "パース成功:\n%s\n" (render_to_string value 0)
    | Error err ->
        Printf.printf "パースエラー: %s\n" (match err with
          | UnexpectedEOF -> "予期しないEOF"
          | InvalidValue msg -> "不正な値: " ^ msg
          | UnclosedString -> "文字列が閉じられていません"
          | UnclosedBlockComment -> "ブロックコメントが閉じられていません"
          | ExpectedChar ch -> "'" ^ String.make 1 ch ^ "' が必要です"
          | InvalidNumber str -> "不正な数値: " ^ str);
    Printf.printf "\n"
  ) test_cases